File: src/ios/UI/ThemeTokens.swift
import SwiftUI

struct ThemeTokens: Identifiable, Sendable {
    let id: String
    let title: String
    let paid: Bool
    let light: Bool
    let background: UInt32
    let surface: UInt32
    let text: UInt32
    let secondary: UInt32
    let accent: UInt32
    let onAccent: UInt32
    let error: UInt32

    static let all = [
        ThemeTokens(id: "afterhours", title: "Afterhours", paid: false, light: false, background: 0x16161A, surface: 0x1F1F25, text: 0xF1EFE8, secondary: 0xB4B2A9, accent: 0x5DCAA5, onAccent: 0x092D23, error: 0xF09595),
        ThemeTokens(id: "daylight", title: "Daylight", paid: false, light: true, background: 0xFAF9F6, surface: 0xFFFFFF, text: 0x22232A, secondary: 0x55565D, accent: 0x176B50, onAccent: 0xFFFFFF, error: 0x9B242C),
        ThemeTokens(id: "ember", title: "Ember", paid: true, light: false, background: 0x211A18, surface: 0x302521, text: 0xF8EDDE, secondary: 0xCCB7A5, accent: 0xEAB375, onAccent: 0x35210D, error: 0xFFAAA2),
        ThemeTokens(id: "lagoon", title: "Lagoon", paid: true, light: false, background: 0x102127, surface: 0x1A3037, text: 0xE3F2EF, secondary: 0xAEC8CD, accent: 0x82D8CC, onAccent: 0x103C37, error: 0xFFAAA2),
        ThemeTokens(id: "violet", title: "Violet", paid: true, light: false, background: 0x1E192A, surface: 0x2B243B, text: 0xF2EBFC, secondary: 0xC3B5D8, accent: 0xD0B2F3, onAccent: 0x352044, error: 0xFFAAA2),
    ]
    static func selected(_ id: String, active: Bool) -> ThemeTokens {
        all.first { $0.id == id && (!$0.paid || active) } ?? all[0]
    }
    static func color(_ rgb: UInt32) -> Color {
        Color(red: Double((rgb >> 16) & 255) / 255, green: Double((rgb >> 8) & 255) / 255, blue: Double(rgb & 255) / 255)
    }
    private static func luminance(_ rgb: UInt32) -> Double {
        func component(_ shift: UInt32) -> Double {
            let c = Double((rgb >> shift) & 255) / 255
            return c <= 0.04045 ? c / 12.92 : pow((c + 0.055) / 1.055, 2.4)
        }
        return component(16) * 0.2126 + component(8) * 0.7152 + component(0) * 0.0722
    }
    static func contrast(_ a: UInt32, _ b: UInt32) -> Double {
        let x = luminance(a), y = luminance(b)
        return (max(x, y) + 0.05) / (min(x, y) + 0.05)
    }
    func readableNickname(_ rgb: UInt32) -> UInt32 {
        Self.contrast(rgb, surface) >= 4.5 && Self.contrast(rgb, background) >= 4.5 ? rgb : text
    }
}

File: src/android/ui/src/main/kotlin/org/meshchat/ui/ThemeTokens.kt
package org.meshchat.ui

import kotlin.math.pow

/** Original palettes. Trust labels never use a sender-supplied color. */
data class ThemeTokens(val id:String,val title:String,val paid:Boolean,val light:Boolean,
    val background:Long,val surface:Long,val text:Long,val secondary:Long,val accent:Long,val onAccent:Long,val error:Long)

object Themes {
    val all=listOf(
        ThemeTokens("afterhours","Afterhours",false,false,0xFF16161A,0xFF1F1F25,0xFFF1EFE8,0xFFB4B2A9,0xFF5DCAA5,0xFF092D23,0xFFF09595),
        ThemeTokens("daylight","Daylight",false,true,0xFFFAF9F6,0xFFFFFFFF,0xFF22232A,0xFF55565D,0xFF176B50,0xFFFFFFFF,0xFF9B242C),
        ThemeTokens("ember","Ember",true,false,0xFF211A18,0xFF302521,0xFFF8EDDE,0xFFCCB7A5,0xFFEAB375,0xFF35210D,0xFFFFAAA2),
        ThemeTokens("lagoon","Lagoon",true,false,0xFF102127,0xFF1A3037,0xFFE3F2EF,0xFFAEC8CD,0xFF82D8CC,0xFF103C37,0xFFFFAAA2),
        ThemeTokens("violet","Violet",true,false,0xFF1E192A,0xFF2B243B,0xFFF2EBFC,0xFFC3B5D8,0xFFD0B2F3,0xFF352044,0xFFFFAAA2),
    )
    fun selected(id:String,active:Boolean):ThemeTokens=all.firstOrNull {it.id==id && (!it.paid || active)} ?: all.first()
    private fun luminance(rgb:Long):Double {
        fun component(shift:Int):Double {
            val c=((rgb shr shift) and 255).toDouble()/255
            return if(c<=0.04045)c/12.92 else ((c+0.055)/1.055).pow(2.4)
        }
        return component(16)*0.2126+component(8)*0.7152+component(0)*0.0722
    }
    fun contrast(a:Long,b:Long):Double {
        val x=luminance(a);val y=luminance(b)
        return (maxOf(x,y)+0.05)/(minOf(x,y)+0.05)
    }
    fun readableNickname(rgb:Long,tokens:ThemeTokens):Long = if(
        contrast(rgb,tokens.surface)>=4.5 && contrast(rgb,tokens.background)>=4.5)rgb else tokens.text
}
